### DISCUSSION TITLE
Include all children in view?

### DISCUSSION THREAD

**Discussion by tombohub** [likec4/likec4#1114] [Q&A] [answered]
I didn't catch in documentation. 

Is it possible to include all children in view?

Given this code 

```javascript
specification {
  element actor
  element system
  element component
}

model {
  customer = actor 'Customer' {
    description 'The regular customer of the system'
  }

  saas = system 'Our SaaS' {
    component ui {
      component admin 
      component frontend
    }
    component backend 'Backend Services' {
      description '
        Implements business logic
        and exposes as REST API
      '
    }

    // UI requests data from the Backend
    ui -> backend 'fetches via HTTPS'
  }

  // Customer uses the UI
  customer -> ui 'opens in browser'
  customer -> saas 'enjoys our product'
}

views {
  view index {
    include *,
    saas.*,
    saas.ui.*,
    saas.backend.*,
  }
}
```

which looks like this:

![image](https://github.com/user-attachments/assets/ab4ab457-297b-41dc-8f15-69008183290b)

Is it possible to replace this:

```javascript
view index {
    include *,
    saas.*,
    saas.ui.*,
    saas.backend.*,
  }
```

to automatically include all `saas` children and grandchildren?

Something like `include saas.all()` ?

**Comment by davydkov** [comment]
Hey @tombohub,

For now there is no predicate to include all descendants

**Reply by Cyclonit** [reply]
I've been thinking about possible options for such a predicate. Glob patterns use `**` for this purpose. Thus `saas.**` would include all descendants of `saas`. Is there any technical reason for why this should not exist? If not, I'd like to give implementing it a shot.

**Reply by davydkov** [reply]
Yes, `.**` seems good to me, and quite clear.

Would be great if you can implement it 🤩 
To minimize conflicts I would suggest you to start from `deployments` branch.

Maybe a good idea try to change
https://github.com/likec4/likec4/blob/3c350b44f3602eaa3aaede2d57ff0e725b3d5c22/packages/core/src/types/expression.ts#L26-L29
to 
```ts
 export interface ElementRefExpr extends Omit<BaseExpr, 'element' | 'isDescedants'> { 
   element: Fqn 
   isDescedants?: boolean | 'all' // or 'glob'
 } 
```
Impact will be minimal, as `all` is also truthy

And here
https://github.com/likec4/likec4/blob/34d43db102a98d4b27739cf63d16fb0367a6f199/packages/core/src/compute-view/element-view/predicates.ts#L28-L29

you may change to (or something better):
```ts
 expr.isDescedants === 'all' ?
   this.graph.descendants(expr.element) : 
   expr.isDescedants === true ? this.graph.childrenOrElement(expr.element) //...
```

**Reply by davydkov** [reply, marked as answer]
#1259 is introducing new predicate `include element.**`
